source/fuzz/transformation_load.cpp

```diff
@@ -61,15 +66,94 @@ bool TransformationLoad::IsApplicable(
       break;
   }
 
+  if (message_.is_atomic()) {
+    // Check the exists of memory scope and memory semantics ids.
+    auto memory_scope_instruction =
+        ir_context->get_def_use_mgr()->GetDef(message_.memory_scope_id());
+    auto memory_semantics_instruction =
+        ir_context->get_def_use_mgr()->GetDef(message_.memory_semantics_id());
+
+    if (!memory_scope_instruction) {
+      return false;
+    }
+    if (!memory_semantics_instruction) {
+      return false;
+    }
+    // The memory scope and memory semantics instructions must have the
+    // 'OpConstant' opcode.
+    if (memory_scope_instruction->opcode() != SpvOpConstant) {
+      return false;
+    }
+    if (memory_semantics_instruction->opcode() != SpvOpConstant) {
+      return false;
+    }
+
+    // The memory scope and memory semantics instructions must have an Integer
+    // operand type with signedness does not matters.
+    if (ir_context->get_def_use_mgr()
+            ->GetDef(memory_scope_instruction->type_id())
+            ->opcode() != SpvOpTypeInt) {
+      return false;
+    }
+    if (ir_context->get_def_use_mgr()
+            ->GetDef(memory_semantics_instruction->type_id())
+            ->opcode() != SpvOpTypeInt) {
+      return false;
+    }
+
+    // The size of the integer for memory scope and memory semantics
+    // instructions must be equal to 32 bits.
+    auto memory_scope_int_width =
+        ir_context->get_def_use_mgr()
+            ->GetDef(memory_scope_instruction->type_id())
+            ->GetInOperand(0)
+            .words[0];
+    auto memory_semantics_int_width =
+        ir_context->get_def_use_mgr()
+            ->GetDef(memory_semantics_instruction->type_id())
+            ->GetInOperand(0)
+            .words[0];
+
+    if (memory_scope_int_width != 32) {
+      return false;
+    }
+    if (memory_semantics_int_width != 32) {
+      return false;
+    }
+
+    // The memory scope constant value must be that of SpvScopeInvocation.
+    auto memory_scope_const_value =
+        memory_scope_instruction->GetInOperand(0).words[0];
+    if (memory_scope_const_value != SpvScopeInvocation) {
+      return false;
+    }
+
+    // The memory semantics constant value must be either
+    // SpvMemorySemanticsWorkgroupMemoryMask or
+    // SpvMemorySemanticsUniformMemoryMask.
+    auto memory_semantics_const_value =
+        memory_semantics_instruction->GetInOperand(0).words[0];
+    if (memory_semantics_const_value != SpvMemorySemanticsUniformMemoryMask &&
+        memory_semantics_const_value != SpvMemorySemanticsWorkgroupMemoryMask) {
+      return false;
+    }
+  }
```

Comment:
Something that I originally asked for but forgot about in my last few reviews: you should check that memory_scope_id and memory_semantics_id are _available_ before the |insert_before| instruction. It might be easier to move the `FindInstruction` code up. Then use `IdIsAvailableBeforeInstruction` to check this.  And then make sure your tests cover the two additional `return false;` statements you will add.

Reply:
Done.

Reply:
You still have not checked that memory_scope_id and memory_semantics_id are _available_ before the |insert_before| instruction, using IdIsAvailableBeforeInstruction.

Reply:
Done.

Reply:
Sorry, I still don't the checks.

* `memory_scope_id` must be available before `insert_before`; if not, return false.
* `memory_semantics_id` must be available before `insert_before`; if not, return false.

To check if an id X is available before an instruction Y, use `fuzzerutil::IdIsAvailableBeforeInstruction(ir_context, Y, X)`.

Reply:
Done. I Hope now is good.